internal/app/handler/handler.go

```diff
@@ -76,3 +80,89 @@ func (handler *Handler) getShortLink(rw http.ResponseWriter, req *http.Request)
 
 	http.Redirect(rw, req, longURL, http.StatusTemporaryRedirect)
 }
+
+// POST /api/shorten
+func (handler *Handler) postAPIShorten(rw http.ResponseWriter, req *http.Request) {
+	type Request struct {
+		URL string `json:"url" valid:"required"`
+	}
+	type Response struct {
+		Result string `json:"result"`
+	}
+
+	var request Request
+	if err := json.NewDecoder(req.Body).Decode(&request); err != nil {
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	valid, err := govalidator.ValidateStruct(request)
+	if err != nil || !valid {
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	shortURL, err := handler.shorten(request.URL)
+	if err != nil {
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	rw.Header().Set("Content-Type", "application/json; charset=utf-8")
+	rw.WriteHeader(http.StatusCreated)
+
+	enc := json.NewEncoder(rw)
+	enc.SetEscapeHTML(false)
+	err = enc.Encode(Response{Result: shortURL})
+	if err != nil {
+		http.Error(rw, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (handler *Handler) shorten(longURL string) (string, error) {
+	shortURL := url.URL{
+		Scheme: handler.baseURL.Scheme,
+		Host:   handler.baseURL.Host,
+	}
+
+	err := urlshortener.ShortenURL(handler.urlShortener, longURL, &shortURL)
```

Comment:
как уже писал про urlshortener.ShortenURL, давай доведем что бы мы могли явно вызвать handler.urlShortener.EncodeLongURL допустим:
```suggestion
	shortURL.Path, err := handler.urlShortener.EncodeLongURL(longURL)
```
Я считаю выглядит гораздо лучше.

Reply:
поправил